tests/unit/test_charm.py

```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
-    @patch("charm.urlopen")
-    def test_add_mongodb_org_repository_success(self, urlopen, from_repo_line, get_mapping):
+    @patch("charm.apt.DebianRepository.import_key")
+    def test_add_mongodb_org_repository_success(self, import_key, from_repo_line, repo_map):
+        repo_map.return_value = set()
+
         self.harness.charm._add_mongodb_org_repository()
-        get_mapping.return_value = {}
-        urlopen.assert_called()
         from_repo_line.assert_called()
+        (from_repo_line.return_value.import_key).assert_called_with(MONGODB_PUB_KEY)
+        self.assertEqual(repo_map.return_value, {from_repo_line.return_value})
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.from_repo_line")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.urlopen")
-    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(self, urlopen, get_mapping):
-        get_mapping.return_value = {}
+    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(
+        self, urlopen, import_key, from_repo_line, repo_map
+    ):
+        # preset values
+        repo_map.return_value = set()
         urlopen.side_effect = URLError("urlopen error")
-        with self.assertLogs("charm", "ERROR") as logs:
-            self.harness.charm._add_mongodb_org_repository()
-            self.assertIn("ERROR:charm:failed to get GPG key, reason:", "".join(logs.output))
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(repo_map.return_value, set())
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't install MongoDB"))
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.apt.DebianRepository.from_repo_line")
     @patch("charm.urlopen")
     def test_add_mongodb_org_repository_cant_create_list_file_blocks(
-        self, urlopen, from_repo_line, get_mapping
+        self, urlopen, from_repo_line, import_key, repo_map
     ):
-        get_mapping.return_value = {}
+        repo_map.return_value = set()
         exceptions = [
             apt.InvalidSourceError("invalid source message"),
             ValueError("value message"),
         ]
-        log_messages = [
-            "ERROR:charm:failed to add repository, invalid source: invalid source message",
-            "ERROR:charm:failed to add repository: value message",
-        ]
-        for exception, log_message in zip(exceptions, log_messages):
-            with self.assertLogs("charm", "ERROR") as logs:
-                from_repo_line.side_effect = exception
-                self.harness.charm._add_mongodb_org_repository()
-                self.assertIn(log_message, "".join(logs.output))
+        for exception in exceptions:
```

Comment:
What is the desired behavior of `_add_mongodb_org_repository()` when it errors with a different error to these?                                                                                                                                                                           I wonder if changing `src/charm.py` to this might help:                                                                                      

```python
try:
    line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
    repo.import_key(key)
    return
except (apt.InavlidSourceError, ValueError):
    logger.exception("failed to add repository due to XYZ reason")  # change XYZ
finally:
    self.unit.status = BlockedStatus("couldn't install MongoDB")
```

That way, in your test you can raise a generic exception, and assert `BlockedStatus`

Reply:
That makes sense! :D

Reply:
@MiaAltieri - Juan caught a bug in my suggestion, I edited the above suggestion for when you want it.

Reply:
Yeah! I saw that with the `finally` as well, I think we aren't supposed to keep the blank `except` either so I think the final form should be:
```
try:
    line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
    repo.import_key(key)
    return
except (apt.InavlidSourceError, ValueError):
    logger.exception("failed to add repository due to XYZ reason")  # change XYZ
    self.unit.status = BlockedStatus("couldn't install MongoDB")
```
Thanks for the help with this! I'll tag you in the next PR :)